[Site Isolation] Color pickers are presented at the wrong location

pxlcoder opened the pull request:
#### f4cc8ce25d0fffddd31fb410aba467bf72412d51
<pre>
[Site Isolation] Color pickers are presented at the wrong location
<a href="https://bugs.webkit.org/show_bug.cgi?id=300117">https://bugs.webkit.org/show_bug.cgi?id=300117</a>
<a href="https://rdar.apple.com/161903415">rdar://161903415</a>

Reviewed by Tim Horton, Abrar Rahman Protyasha, and Lily Spiniolas.

iframe processes cannot convert to root view coordinates, resulting in the
popover being presented at the wrong location.

Fix by converting the rect in the UI process.

* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::rootFrameID const):
* Source/WebCore/html/ColorInputType.h:
* Source/WebCore/platform/ColorChooserClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::showColorPicker):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebColorChooser.cpp:
(WebKit::WebColorChooser::WebColorChooser):
(WebKit::WebColorChooser::reattachColorChooser):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, ColorInputPickerLocation)):

Canonical link: <a href="https://commits.webkit.org/301062@main">https://commits.webkit.org/301062@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52ed14c7eb95985b415966f4bffed61e1e7ecf14

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124505 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44183 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34917 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131343 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76484 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/41ce8ce7-3955-41cd-a3c1-a92ef3085d85) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126382 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44900 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52764 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94709 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62815 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/09afe00a-c581-44d0-86e7-24a5056fdecc) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127459 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35787 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111349 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75287 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cc22ff96-0de5-4b35-8b7c-7f7eb9f78ee5) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34734 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29505 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74824 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105539 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29736 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134008 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51375 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39201 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103191 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51783 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107571 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102973 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48339 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26599 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48318 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19563 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51246 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57036 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50656 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54015 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52336 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/95251852bb7ce2d23650ea5b6b102e720f099862. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51762

aprotyas commented on Source/WebKit/WebProcess/WebCoreSupport/WebColorChooser.cpp:
I don't think the frame ID copy gets elided here, consider moving it?

pxlcoder replied:
GTK bots showed this: `static_assert(is_lvalue_reference<T>::value, "T is not an lvalue reference; move() is unnecessary.");`

I removed the move, but kept as an rvalue in `WebPageProxy::showColorPicker`.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/71c233fc2cdb808ae5e797a9c771891e3f1cd29a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51762

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/52ed14c7eb95985b415966f4bffed61e1e7ecf14. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51762

pxlcoder committed:
[Site Isolation] Color pickers are presented at the wrong location
https://bugs.webkit.org/show_bug.cgi?id=300117
rdar://161903415

Reviewed by Tim Horton, Abrar Rahman Protyasha, and Lily Spiniolas.

iframe processes cannot convert to root view coordinates, resulting in the
popover being presented at the wrong location.

Fix by converting the rect in the UI process.

* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::rootFrameID const):
* Source/WebCore/html/ColorInputType.h:
* Source/WebCore/platform/ColorChooserClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::showColorPicker):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebColorChooser.cpp:
(WebKit::WebColorChooser::WebColorChooser):
(WebKit::WebColorChooser::reattachColorChooser):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, ColorInputPickerLocation)):

Canonical link: https://commits.webkit.org/301062@main
